test/parallel/test-async-wrap-propagate-parent.js

```diff
@@ -4,40 +4,92 @@ const common = require('../common');
 const assert = require('assert');
 const net = require('net');
 const async_wrap = process.binding('async_wrap');
+const providers = Object.keys(async_wrap.Providers);
+
+Error.stackTraceLimit = Infinity;
+process._rawDebug('# providers: ', async_wrap.Providers);
 
 let cntr = 0;
-let server;
+let serverUid = -1;
+let serverHandle;
 let client;
 
-function init(type, id, parent) {
-  if (parent) {
+function init(uid, type, parentUid, parentHandle) {
+  const handle = this;
+
+  process._rawDebug('init: ', {
+    name: this.constructor.name,
+    uid, type, parentUid, parentHandle
+  });
+  process._rawDebug('stack: ', new Error('trace').stack);
+
+  // The first TCPWRAP is the server
+  if (type === async_wrap.Providers.TCPWRAP && serverUid === -1) {
+    process._rawDebug('# set server uid and handle');
+    serverUid = uid;
+    serverHandle = this;
+  }
+
+  else if (parentUid) {
     cntr++;
+
+    process.nextTick(function () {
+      process._rawDebug('assert: ', {
+        provider: providers[type],
+        uid: {parentUid, serverUid},
+        serverHandle: parentHandle === serverHandle,
+        clientHandle: handle === client._handle
+      });
+    });
+
+    /*
     // Cannot assert in init callback or will abort.
     process.nextTick(() => {
-      assert.equal(parent, server._handle, 'server doesn\'t match parent');
+      assert.equal(providers[type], 'TCPWRAP');
+      assert.equal(parentUid, serverUid,
+                   'server uid doesn\'t match parent uid');
+      assert.equal(parentHandle, serverHandle,
+                   'server handle doesn\'t match parent handle');
       assert.equal(this, client._handle, 'client doesn\'t match context');
     });
+    */
   }
 }
 
-function noop() { }
+function pre(uid) {
+  process._rawDebug('# pre: ' + uid);
+}
+
+function post(uid) {
+  process._rawDebug('# post: ' + uid);
+}
 
-async_wrap.setupHooks(init, noop, noop);
+async_wrap.setupHooks(init, pre, post);
+process._rawDebug('# enable');
 async_wrap.enable();
 
-server = net.createServer(function(c) {
+process._rawDebug('# common.PORT: ' + common.PORT);
+
+const server = net.createServer(function(c) {
+  process._rawDebug('# on connection');
+
   client = c;
   // Allow init callback to run before closing.
-  setImmediate(() => {
+  setImmediate(function () {
+    process._rawDebug('# stop server');
     c.end();
-    this.close();
+    server.close();
   });
 }).listen(common.PORT, function() {
-  net.connect(common.PORT, noop);
+  process._rawDebug('# listening on port: ' + this.address().port);
+  net.connect(common.PORT, function () { });
 });
 
 
 process.on('exit', function() {
+  process._rawDebug('# exit');
+
   // init should have only been called once with a parent.
   assert.equal(cntr, 1);
+  assert(false);
```

Comment:
We are currently debugging some odd issue that only happens on raspberry pi in the test environment. For Jenkins to print the stdout the process has to fail. Once the issue gets resolved I will remove the debug commit that adds the assert (among other things).